README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->

The goalmodel package let you build prediction models for the number of
goals scored in sport games. The models are primarily aimed at modelling
and predicting football (soccer) scores, but could also be applicable
for similar sports, such as hockey and handball.

If you find this package useful, please consider supporting the
development at <https://ko-fi.com/opisthokonta>

[![ko-fi banner and
link](README_files/SupportMe_red@2x_scaled.png)](https://ko-fi.com/opisthokonta)

# Installation

``` r
install.packages("devtools")
devtools::install_github("opisthokonta/goalmodel")
```

# Whats new

### Version 0.6

-   New functions predict_btts() and pbtts() for computing
    Both-Teams-To-Score probabilities.
-   New function score_predictions(), for evaluating prediction accuracy
    using Ranked Probability Scores, Brier scores, and Log scores.
-   New function league_table() that computes a league table with points
    and useful summary statistics.

### Version 0.5

-   New function expg_from_ou() - Finds the expected goals from
    over/under probabilities.
-   New function p1x2() - Compute win and draw probabilities from
    expected goals.
-   Bugfix: predict_goals() with return_df=TRUE did not work properly
    when the team arguments were factors.

See NEWS.md for changes for complete version history.

# The default model

The goalmodel package models the goal scoring intensity so that it is a
function of the two opposing teams attack and defense ratings. Let
λ<sub>1</sub> and λ<sub>2</sub>) be the goal scoring intensities for the
two sides. The default model in the goalmodel package then models these
as

log(λ<sub>1</sub>) = μ + hfa + attack<sub>1</sub> − defense<sub>2</sub>

log(λ<sub>2</sub>) = μ + attack<sub>2</sub> − defense<sub>1</sub>

where μ is the intercept (approximately the average number of goals
scored) and hfa is the home field advantage given to team 1. By default
the number of goals scored, Y<sub>1</sub> and Y<sub>2</sub> are
distributed as

Y<sub>1</sub> ∼ Poisson(λ<sub>1</sub>)

Y<sub>2</sub> ∼ Poisson(λ<sub>2</sub>)

The default model can be modified in numerous ways as detailed in this
vignette.

## Fitting the default model

Models are fitted with the goalmodel function. The minimum required
input are vectors containing the number of goals scored and the names of
the teams. Here I use data from the excellent
[engsoccerdata](https://cran.r-project.org/web/packages/engsoccerdata/)
package.

``` r
library(goalmodel)
library(dplyr) # Useful for data manipulation. 
library(engsoccerdata) # Some data.
library(Rcpp)

# Load data from English Premier League, 2011-12 season.
england %>% 
  filter(Season %in% c(2011),
         tier==c(1)) %>% 
  mutate(Date = as.Date(Date)) -> england_2011


# Fit the default model, with the home team as team1.

gm_res <- goalmodel(goals1 = england_2011$hgoal, goals2 = england_2011$vgoal,
                     team1 = england_2011$home, team2=england_2011$visitor)

# Show the estimated attack and defense ratings and more.
summary(gm_res)
```

    ## Model sucsessfully fitted in 0.03 seconds
    ## 
    ## Number of matches           380 
    ## Number of teams              20 
    ## 
    ## Model                     Poisson 
    ## 
    ## Log Likelihood            -1088.99 
    ## AIC                        2257.98 
    ## R-squared                  0.18 
    ## Parameters (estimated)       40 
    ## Parameters (fixed)            0 
    ## 
    ## Team                      Attack   Defense
    ## Arsenal                    0.36     0.03 
    ## Aston Villa               -0.33    -0.01 
    ## Blackburn Rovers          -0.04    -0.41 
    ## Bolton Wanderers          -0.09    -0.39 
    ## Chelsea                    0.23     0.10 
    ## Everton                   -0.04     0.26 
    ## Fulham                    -0.07     0.02 
    ## Liverpool                 -0.10     0.26 
    ## Manchester City            0.57     0.54 
    ## Manchester United          0.53     0.41 
    ## Newcastle United           0.08     0.01 
    ## Norwich City               0.02    -0.25 
    ## Queens Park Rangers       -0.17    -0.24 
    ## Stoke City                -0.36    -0.01 
    ## Sunderland                -0.14     0.12 
    ## Swansea City              -0.16     0.02 
    ## Tottenham Hotspur          0.24     0.22 
    ## West Bromwich Albion      -0.13     0.00 
    ## Wigan Athletic            -0.19    -0.17 
    ## Wolverhampton Wanderers   -0.22    -0.45 
    ## -------
    ## Intercept                  0.13 
    ## Home field advantage       0.27

# The Dixon-Coles model

In a paper by Dixon and Coles (1997) they introduce a model that extend
the defaul model that modifies the probabilities for low scores, where
each team scores at most 1 goal each. The amount of adjustment is
determined by the parameter ρ (rho), which can be estimated from the
data. The Dixon-Coles model can be fitted by setting the dc argument to
TRUE.

``` r
# Fit the Dixon-Coles model.

gm_res_dc <- goalmodel(goals1 = england_2011$hgoal, goals2 = england_2011$vgoal,
                     team1 = england_2011$home, team2=england_2011$visitor,
                     dc=TRUE)

summary(gm_res_dc)
```

    ## Model sucsessfully fitted in 0.88 seconds
    ## 
    ## Number of matches           380 
    ## Number of teams              20 
    ## 
    ## Model                     Poisson 
    ## 
    ## Log Likelihood            -1087.36 
    ## AIC                        2256.72 
    ## R-squared                  0.18 
    ## Parameters (estimated)       41 
    ## Parameters (fixed)            0 
    ## 
    ## Team                      Attack   Defense
    ## Arsenal                    0.37     0.03 
    ## Aston Villa               -0.31    -0.03 
    ## Blackburn Rovers          -0.06    -0.41 
    ## Bolton Wanderers          -0.08    -0.40 
    ## Chelsea                    0.23     0.09 
    ## Everton                   -0.06     0.27 
    ## Fulham                    -0.07     0.01 
    ## Liverpool                 -0.11     0.25 
    ## Manchester City            0.56     0.55 
    ## Manchester United          0.52     0.43 
    ## Newcastle United           0.10     0.00 
    ## Norwich City               0.02    -0.25 
    ## Queens Park Rangers       -0.18    -0.23 
    ## Stoke City                -0.36    -0.01 
    ## Sunderland                -0.14     0.12 
    ## Swansea City              -0.15     0.01 
    ## Tottenham Hotspur          0.24     0.21 
    ## West Bromwich Albion      -0.14     0.00 
    ## Wigan Athletic            -0.19    -0.17 
    ## Wolverhampton Wanderers   -0.21    -0.46 
    ## -------
    ## Intercept                  0.12 
    ## Home field advantage       0.27 
    ## Dixon-Coles adj. (rho)    -0.13

Notice that the estimated ρ parameter is listed together with the other
parameters.

# The Rue-Salvesen adjustment

In a paper by Rue and Salvesen (2001) they introduce an adjustment to
the goals scoring intesities λ<sub>1</sub> and λ<sub>2</sub>:

log(λ<sub>1</sub><sup>adj</sup>) = log(λ<sub>1</sub>) − γΔ<sub>1,2</sub>

log(λ<sub>2</sub><sup>adj</sup>) = log(λ<sub>2</sub>) + γΔ<sub>1,2</sub>

where

Δ<sub>1,2</sub> =
(attack<sub>1</sub>+defense<sub>1</sub>−attack<sub>2</sub>−defense<sub>2</sub>)/2

and γ is a parameter that modifies the amount of adjustment. This model
can be fitted by setting the rs argument to TRUE.

``` r
# Fit the model with the Rue-Salvesen adjustment.

gm_res_rs <- goalmodel(goals1 = england_2011$hgoal, goals2 = england_2011$vgoal,
                     team1 = england_2011$home, team2=england_2011$visitor,
                     rs=TRUE)

summary(gm_res_rs)
```

    ## Model sucsessfully fitted in 0.54 seconds
    ## 
    ## Number of matches           380 
    ## Number of teams              20 
    ## 
    ## Model                     Poisson 
    ## 
    ## Log Likelihood            -1088.99 
    ## AIC                        2259.98 
    ## R-squared                  0.18 
    ## Parameters (estimated)       41 
    ## Parameters (fixed)            0 
    ## 
    ## Team                      Attack   Defense
    ## Arsenal                    0.36     0.03 
    ## Aston Villa               -0.33    -0.02 
    ## Blackburn Rovers          -0.05    -0.42 
    ## Bolton Wanderers          -0.09    -0.40 
    ## Chelsea                    0.23     0.10 
    ## Everton                   -0.04     0.26 
    ## Fulham                    -0.07     0.01 
    ## Liverpool                 -0.10     0.26 
    ## Manchester City            0.58     0.54 
    ## Manchester United          0.54     0.42 
    ## Newcastle United           0.09     0.01 
    ## Norwich City               0.02    -0.25 
    ## Queens Park Rangers       -0.17    -0.24 
    ## Stoke City                -0.36    -0.02 
    ## Sunderland                -0.14     0.12 
    ## Swansea City              -0.16     0.02 
    ## Tottenham Hotspur          0.24     0.22 
    ## West Bromwich Albion      -0.13    -0.00 
    ## Wigan Athletic            -0.20    -0.18 
    ## Wolverhampton Wanderers   -0.23    -0.46 
    ## -------
    ## Intercept                  0.13 
    ## Home field advantage       0.27 
    ## Rue-Salvesen adj. (gamma)  0.02

# Making predictions

There is little use in fitting the model without making predictions.
Several functions are provided that make different types of predictions.

We can get the expected goals with the predict_expg function. Here the
return_df is set to TRUE, which returns a convenient data.frame.

``` r
to_predict1 <- c('Arsenal', 'Manchester United', 'Bolton Wanderers')
to_predict2 <- c('Fulham', 'Chelsea', 'Liverpool')

predict_expg(gm_res_dc, team1=to_predict1, team2=to_predict2, return_df = TRUE)
```

    ##               team1     team2    expg1     expg2
    ## 1           Arsenal    Fulham 2.119358 1.0243004
    ## 2 Manchester United   Chelsea 2.291612 0.9288743
    ## 3  Bolton Wanderers Liverpool 1.069974 1.5090408

We can also get the probabilities of the final outcome (team1 win, draw,
team2 win) with the predict_result function.

``` r
predict_result(gm_res_dc, team1=to_predict1, team2=to_predict2, return_df = TRUE)
```

    ##               team1     team2        p1        pd        p2
    ## 1           Arsenal    Fulham 0.6121100 0.2266543 0.1612357
    ## 2 Manchester United   Chelsea 0.6685366 0.2051625 0.1263009
    ## 3  Bolton Wanderers Liverpool 0.2524546 0.2903231 0.4572222

Other functions are predict_goals which return matrices of the
probabilities of each scoreline, and predict_ou for getting the
probabilities for over/under total scores.

# Weights

For predicton purposes it is a good idea to weight the influence of each
match on the estimates so that matches far back in time have less
influence than the more recent ones. The Dixon-Coles paper describe a
function that can help set the weights. This function is implemented in
the weights_dc function. The degree more recent matches are weighted
relative to earlier ones are determined by the parameter xi. Good values
of xi is usually somewhere between 0.001 and 0.003

Here we calculate the weights, plot them, and fit the default model with
the weights.

``` r
my_weights <- weights_dc(england_2011$Date, xi=0.0019)

length(my_weights)
```

    ## [1] 380

``` r
plot(england_2011$Date, my_weights)
```

![](README_files/figure-markdown_github/unnamed-chunk-7-1.png)

``` r
gm_res_w <- goalmodel(goals1 = england_2011$hgoal, goals2 = england_2011$vgoal, 
                     team1 = england_2011$home, team2=england_2011$visitor,
                     weights = my_weights)
```

# Additional covariates

It is possible to use additional variables to predict the outcome. One
particular useful thing this can be used for is to have greater control
of the home field advantage. For example if you use data from several
leagues you might want to have a separate home field advantage for each
league, or be able to turn of the advantage for games played on neutral
ground.

Here is how to specify the default model with the home field advantage
specified as an additional covariate. Notice how the hfa argument is set
to FALSE, so that the default home field advantage factor is droped.
Also notice that only covariates for the first team is given, implicitly
setting all values of this covariate to 0 for team2.

``` r
# Create a n-by-1 matrix, containing only 1's, to indicate that all 
# games have a home team.
xx1_hfa <- matrix(1, nrow=nrow(england_2011), ncol=1)
colnames(xx1_hfa) <- 'hfa' # must be named.

# Fit the default model, with the home team as team1.
gm_res_hfa <- goalmodel(goals1 = england_2011$hgoal, goals2 = england_2011$vgoal,
                     team1 = england_2011$home, team2=england_2011$visitor,
                    hfa = FALSE, # Turn of the built-in Home Field Advantage
                    x1 = xx1_hfa) # Provide the covariate matrix.

# the coefficients for the additional covariates are found in the 'beta' element in 
# the parameter list. They will also show up in the summary.
gm_res_hfa$parameters$beta
```

    ##       hfa 
    ## 0.2680093

# Alternatives to the Poisson model

By the default the Poisson distribution is used to model the goals. The
Poisson distribution has a lot of nice properties, but also some
limitations. For instance there is only one parameter that describes
both the expected value (the average) and the variance. It is also
limited to integer number of goals.

## The Negative Binomial model

The Negative Binomial (NB) model is a more flexible model that allows
higher variance than the Poisson model. The NB model introduces another
parameter called the “dispersion” parameter. This parameter has to be
positive, and the greater the value of this parameter, the more variance
there is relative to the Poisson model with the same *λ*. When the
variance is greater than the expectation, it is referred to as
overdispersion. When the parameter is close to zero, it the result is
approximately the Poisson. The NB model can be fitted by setting the
“model” argument to ‘negbin’. In this example the dispersion parameter
is close to zero.

``` r
# Fit the model with overdispersion.

gm_res_nb <- goalmodel(goals1 = england_2011$hgoal, goals2 = england_2011$vgoal,
                     team1 = england_2011$home, team2=england_2011$visitor,
                     model='negbin')


gm_res_nb$parameters$dispersion
```

    ## [1] 4.517136e-05

## The Conway-Maxwell-Poisson model

While the Negative Binomial model is more flexible than the Poisson

NEWS.md


# Version 0.6

 * New functions predict_btts() and pbtts() for computing Both-Teams-To-Score probabilities.
 * New function score_predictions(), for evaluating prediction accuracy using Ranked Probability Scores, Brier scores, and Log scores.
 * New function league_table() that computes a league table with points and useful summary statistics. 


# Version 0.5
 
 * New function expg_from_ou().
 * New function p1x2().
 * Bugfix: predict_goals() with return_df=TRUE did not work properly when the team arguments were factors.



# Version 0.4.3

  * Pseudo R-squared is now computed for fitted CMP models.
  * Can now do weighted estimation of the CMP dispersion parameter with the upsilon.ml() function.
  * New argument maxgoal in predict_goals()
  * predict_expg() does not return a data.frame with row.names if return_df=TRUE.


# Version 0.4.2

  * New function dDCP(), the probability mass function for the Dixon-Coles model.


# goalmodel 0.4.1

  * New function qCMP(), the quantile function for the Conway-Maxwell-Poisson distribution.
  
  * The predict_goals() function has a new argument return_df. If return_df = TRUE, the function returns a long data.frame instead of a list of matrices. 

# goalmodel 0.4

 * The prediction functions now works when trying to predict the outcome when a team is not in the model. The predictions will be NA.

  * Some of the functions related to the Conway-Maxwell-Poisson distribution is re-implemented in Rcpp. This should give somewhat faster estimation.

  * Bugfix: The returned loglikelihood (and R squared, aic, etc) when using weights were sometimes wrong. This is now fixed.



# goalmodel 0.3

  * New function expg_from_probabilities(). This converts win-draw-lose probabilities to expected goals.

  * Two new functions matches_last_xdays() and days_since_last_match(). These are useful if you want to include information about the effect of having a busy schedule into your model.

  * Some additional checks have been added make sure that the network of all team that has played each other is fully connected.



# goalmodel 0.2

   * New faster model fitting. Some models use the built-in glm.fit function to estimate the attack and defence parameters. If the model can not be estimated with the glm.fit function, the results from glm.fit is used as starting values for subsequent fitting. All this should give faster estimation.

   * Can now fit Conway-Maxwell-Poisson (CMP) models. This is still a bit slow and unstable, so two-step estimation is recomended. See the section on the CMP below for more info. Several new functions related to the CMP model is also available if you want to use them for additional modelling.

   * The model = 'ls' option is removed. The option to use model = 'gaussian', is still available and should be equivalent to the old 'ls' option.

   * Some changes to what is returned by the goalmodel function, but this should not cause much problems to your existing code since it is only related to some of the more technical parts of the output.

# goalmodel 0.1
   * Initial release



R/goalmodel_misc.R
# This script contains some miscellaneous functions, not directly invovled
# in model fitting and making predictions.

#' Print a summary of a fitted goalmodel.
#'
#' @param object A fitted goalmodel object.
#' @param ... ignored.
#'
#' @export
summary.goalmodel <- function(object, ...){

  if (length(object$warnings) == 0){
    cat(sprintf('Model sucsessfully fitted in %.2f seconds\n\n', object$est_time))
  } else {
    cat(sprintf('Model fitted with warnings (%d). Results may be unreliable.\n\n', length(object$warnings)))
  }

  # Data summary
  cat(sprintf('%-25.25s % 5.0d \n', 'Number of matches', object$ngames))
  cat(sprintf('%-25.25s % 5.0d \n', 'Number of teams', length(object$all_teams)))

  cat('\n')

  if (object$model == 'poisson'){
    cat(sprintf('%-25.25s %s \n', 'Model', 'Poisson'))
  } else if (object$model == 'negbin'){
    cat(sprintf('%-25.25s %s \n', 'Model', 'Neg. Binom.'))
  } else if (object$model == 'gaussian'){
    cat(sprintf('%-25.25s %s \n', 'Model', 'Gaussian'))
  } else if (object$model == 'cmp'){
    cat(sprintf('%-25.25s %s \n', 'Model', 'CMP'))
  }

  cat('\n')

  # fit indices
  cat(sprintf('%-25.25s % 2.2f \n', 'Log Likelihood', object$loglikelihood))
  cat(sprintf('%-25.25s % 2.2f \n', 'AIC', object$aic))
  cat(sprintf('%-25.25s % 2.2f \n', 'R-squared', object$r_squared))
  cat(sprintf('%-25.25s % 5.1d \n', 'Parameters (estimated)', object$npar_est))
  cat(sprintf('%-25.25s % 5.1d \n', 'Parameters (fixed)', object$npar_fixed))

  cat('\n')

  # Attack and defense parameters
  cat(sprintf('%-25s %-3s   %-3s\n', 'Team', 'Attack', 'Defense'))
  for (ii in 1:length(object$all_teams)){
    ct <- object$all_teams[ii]
    cat(sprintf('%-25.25s % 2.2f    % 2.2f \n',
                ct,
                object$parameters$attack[ct],
                object$parameters$defense[ct]))
  }
  cat('-------\n')
  cat(sprintf('%-25.25s % 2.2f \n', 'Intercept', object$parameters$intercept))

  if ('hfa' %in% names(object$parameters)){
    cat(sprintf('%-25.25s % 2.2f \n', 'Home field advantage', object$parameters$hfa))
  }

  if ('rho' %in% names(object$parameters)){
    cat(sprintf('%-25.25s % 2.2f \n', 'Dixon-Coles adj. (rho)', object$parameters$rho))
  }

  if ('dispersion' %in% names(object$parameters) & object$model == 'negbin'){
    cat(sprintf('%-25.25s % 2.2f \n', 'Dispersion (Neg. Binom.)', object$parameters$dispersion))
  }

  if ('dispersion' %in% names(object$parameters) & object$model == 'cmp'){
    cat(sprintf('%-25.25s % 2.2f \n', 'Dispersion (CMP)', object$parameters$dispersion))
  }


  if ('sigma' %in% names(object$parameters)){
    cat(sprintf('%-25.25s % 2.2f \n', 'Sigma (Gaussian)', object$parameters$sigma))
  }

  if ('gamma' %in% names(object$parameters)){
    cat(sprintf('%-25.25s % 2.2f \n', 'Rue-Salvesen adj. (gamma)', object$parameters$gamma))
  }

  if ('beta' %in% names(object$parameters)){
    for (ii in 1:length(object$parameters$beta)){
      cat(sprintf('%-25.25s % 2.2f \n', names(object$parameters$beta)[ii], object$parameters$beta[ii]))
    }
  }

}


#' Compute Dixon-Coles weights
#'
#' This function computes weights that give less weight to games further back in time.
#'
#' The weigths w are computed using the formula
#'
#' w = exp(-xi*t)
#'
#' where xi is a tuning parameter that decides the amount of down-weighting,
#' and t is the number of days since currentDate. If any date in dates are after
#' currentDate, the corresponding weights are set to 0.
#'
#'
#' @param dates a vector of dates (of type Date).
#' @param xi A numeric with the time dumping factor. Ususally a value between 0.001 and 0.003.
#' @param currentDate The date which to count backwards from. Default to the latest date in dates.
#'
#' @export
weights_dc <- function(dates, xi=0, currentDate=NULL){
  stopifnot(xi >= 0,
            length(xi) == 1)
  dates <- as.Date(dates)

  if (is.null(currentDate)){
    currentDate <- max(dates)
  } else {
    currentDate <- as.Date(currentDate)
  }

  datediffs <- dates - as.Date(currentDate)
  datediffs <- as.numeric(datediffs *-1)
  w <- exp(-1*xi*datediffs)
  w[datediffs < 0] <- 0 #Future dates should have zero weights
  return(w)
}



# Internal function for computing the squared error for estimating
# the expected goals from outcome probabilities.
expg_prob_sq_error <- function(pars, trgt_probs, rho, uprx){

  pars <- exp(pars) # trick to avoid negaive lambda parameters.
  hda_probs <- numeric(3)
  probmat <- stats::dpois(0:uprx, lambda=pars[1]) %o% stats::dpois(0:uprx, lambda=pars[2])

  ## DC adjustment
  if (rho != 0){
    correctionmat <- matrix(tau(c(0,1,0,1), c(0,0,1,1),
                                rep(pars[1], 4),
                                rep(pars[2], 4), rho), nrow=2)
    probmat[1:2, 1:2] <- probmat[1:2, 1:2] * correctionmat
  }

  hda_probs[2] <- sum(diag(probmat))
  hda_probs[1] <- sum(probmat[lower.tri(probmat)])
  hda_probs[3] <- 1 - sum(hda_probs[1:2])

  sum((hda_probs - trgt_probs)^2)

}

#' Estimate the expected goals from win-draw-lose probabilities.
#'
#' This function converts outcome probabilities into expected goals,
#' assuming an underlying Poisson distribution, or a Dixon-Coles-Poisson
#' distribution with known dependence paramater rho.
#'
#' @param probabilities A 3-column matrix of win-draw-lose probabilities.
#' @param rho numeric. Value for the Dixon-Coles adjustment parameter. Default is 0, which is the same as no adjustment.
#' @param uprx numeric. The upper limit for evaluating the poisson distribution.
#'
#' @return A list with two elements. The first, expg, is a two-column matrix of
#' expected goals. The second, sq_error, is a numeric vector indicating the how well
#' the expected goals matches the probabilities using the poisson distribution.
#'
#'
#' @export
expg_from_probabilities <- function(probabilities, rho=0, uprx=75){

  # Convert to matrix
  if (!is.matrix(probabilities)){
    if (is.numeric(probabilities)){
      probabilities <- matrix(probabilities, nrow=1,
                                 dimnames = list(NULL, names(probabilities)))
    } else {
      probabilities <- as.matrix(probabilities)
    }


  }

  stopifnot(ncol(probabilities) == 3,
            all(abs(rowSums(probabilities) - 1) < 0.0001),
            uprx >= 1,
            length(uprx) == 1,
            length(rho) == 1)

  expg <- matrix(ncol=2, nrow=nrow(probabilities))
  sq_errors <- numeric(nrow(probabilities))

  for (ii in 1:nrow(probabilities)){

    optim_res <- stats::optim(c(0,0), fn=expg_prob_sq_error,
                       trgt_prob=probabilities[ii,],
                       rho = rho, uprx = uprx)

    expg[ii,] <- exp(optim_res$par)
    sq_errors[ii] <- optim_res$value

  }

  out <- list(expg = expg, sq_errors=sq_errors)

  return(out)
}



#' Estimate the expected goals from over/under probabilities.
#' From a probability of goals scored being below a given number (like 2.5), compute
#' the expected number of goals scored, assuming an underlying Poisson distribution.
#'
#' @param probability Vector of probabilities for number of goals scored being less (or greater) than ou.
#' @param ou The limit the probability refers to. Default is 2.5.
#' @param under Logical indicating whether the probabilities are for 'under' (TRUE, default) or 'over' (FALSE).
#' @param upperlim Numeric giving the upper limit for expected goals the estimating procedure will consider.
#'
#'@return
#' A numeric vector with expected number of goals.
#'
#' @export
expg_from_ou <- function(probability, ou=2.5, under = TRUE, upperlim = 10){

  stopifnot(all(probability >= 0),
            all(probability <= 1),
            length(ou) == 1,
            ou > 0,
            is.logical(under),
            length(under) == 1)

  # If probabilities are given as 'over', convert them to 'under'.
  if (!under){
    probability <- 1 - probability
  }

  obj <- function(expg, prob, under){
    stats::ppois(floor(under), lambda=expg) - prob
  }

  expg <- numeric(length(probability))
  for (ii in 1:length(probability)){

    res <- stats::uniroot(f = obj,
                         interval = c(0.001, upperlim),
                         prob = probability[ii],
                         under = ou)

    expg[ii] <- res$root
  }

  return(expg)

}



#' Time since last match
#'
#' This function calculates the number of days since the two teams
#' played their respective previous matches.
#'
#' @param team1 Vector of team names.
#' @param team2 Vector of team names.
#' @param dates a vector of dates (of type Date).
#' @param first_val the value used for a team's first match in the data, when there
#' are no preceding matches. Default is NA.
#'
#' @return A two-column matrix with the number of days since last time
#' each team played.
#'
#' @seealso \code{\link{matches_last_xdays}}
#'
#' @export
days_since_last_match <- function(team1, team2, dates, first_val = NA){

  stopifnot(length(team1) == length(team2),
            length(team2) == length(dates),
            length(team1) >= 1,
            length(first_val) == 1)

  dates <- as.Date(dates)
  ngames <- length(team1)
  all_teams <- sort(unique(c(unique(team1), unique(team2))), decreasing = FALSE)

  res <- matrix(NA, ncol=2, nrow=ngames)

  for (ii in 1:length(all_teams)){

    team_idx1 <- team1 == all_teams[ii]
    team_idx2 <- team2 == all_teams[ii]

    tidx <- as.numeric(team_idx1)
    tidx[team_idx2] <- 2

    team_idx <- which(team_idx1 | team_idx2)

    tdates <- dates[team_idx]

    for (tt in 1:length(tdates)){
      if (tdates[tt] == min(tdates)){
        tmp_val <- first_val
      } else {
        tmp_val <- tdates[tt] - max(tdates[tdates < tdates[tt]])
      }

      res[team_idx[tt], tidx[team_idx[tt]]] <- tmp_val

    }
  }

  return(res)

}


#' Number of matches played in the preceding period
#'
#' This function calculates the number of matches the two teams
#' have played the last x days.
#'
#' @param team1 Vector of team names.
#' @param team2 Vector of team names.
#' @param dates a vector of dates (of type Date).
#' @param days_since The number of days back in time to count matches.
#' @param first_val the value used for a team's first match in the data, when there
#' are no preceding matches. Default is NA.
#'
#' @return A two-column matrix with the number of matches played in the preceding
#' period of time, as given by the days_since argument.
#'
#' @seealso \code{\link{days_since_last_match}}
#'
#' @export
matches_last_xdays <- function(team1, team2, dates, days_since=30, first_val = NA){

  stopifnot(length(team1) == length(team2),
            length(team2) == length(dates),
            length(team1) >= 1,
            length(days_since) == 1,
            is.numeric(days_since))

  dates <- as.Date(dates)
  ngames <- length(team1)
  all_teams <- sort(unique(c(unique(team1), unique(team2))), decreasing = FALSE)

  res <- matrix(NA, ncol=2, nrow=ngames)

  for (ii in 1:length(all_teams)){

    team_idx1 <- team1 == all_teams[ii]
    team_idx2 <- team2 == all_teams[ii]

    tidx <- as.numeric(team_idx1)
    tidx[team_idx2] <- 2

    team_idx <- which(team_idx1 | team_idx2)

    tdates <- dates[team_idx]

    for (tt in 1:length(tdates)){
      if(tdates[tt] == min(tdates)){
        res[team_idx[tt], tidx[team_idx[tt]]] <- first_val
      } else {
        tmp_val <- pmax((tdates[tt] - tdates),0)
        res[team_idx[tt], tidx[team_idx[tt]]] <- sum(tmp_val <= days_since & tmp_val > 0)
      }

    }
  }

  return(res)

}





#' Create league table
#'
#' Creates a league table with basic descriptive statistics and number of points.
#'
#' Points are calculated as 3 points for win, 1 for draw, and 0 for loss. Results
#' are sorted by points and ties will be resolved by 1) goal difference and 2) goals scored.
#' Further ties will not be resolved, and the tied teams will be ranked arbitrarily.
#'
#' Note that leagues and competitions will often have different rules for how teams
#' are ranked, especially for how ties are handled. The results provided by this functions
#' may therefore differ from official league rankings.
#'
#' Games with missing values in the number of goals scores will be ignored and will
#' therefore not contribute to the table.
#'
#' @param goals1 Numeric, non-negative integer. The number of goals scored by team 1.
#' @param goals2 Numeric, non-negative integer. The number of goals scored by team 2.
#' @param team1 Vector of team names.
#' @param team2 Vector of team names.
#'
#' #' @return
#'  A data.frame with one row for each team in the data.
#'
#' @export
league_table <- function(goals1, goals2, team1, team2){

  team1 <- as.character(team1)
  team2 <- as.character(team2)

  stopifnot(length(goals1) == length(goals2),
            length(goals2) == length(team1),
            length(team1) == length(team2),
            is.numeric(goals1),
            is.numeric(goals2))

  #Remove missing values
  mising_idx <- is.na(goals1) | is.na(goals2) | is.na(team1) | is.na(team2)
  goals1 <- goals1[!mising_idx]
  goals2 <- goals2[!mising_idx]
  team1 <- team1[!mising_idx]
  team2 <- team2[!mising_idx]

  all_teams <- sort(unique(c(unique(team1), unique(team2))), decreasing = FALSE)
  n_teams <- length(all_teams)
  ngames <- length(goals1)

  goaldiff <- goals1 - goals2


  # init result vectors
  games_played <- rep(0, n_teams)
  names(games_played) <- all_teams

  points <- rep(0, n_teams)
  names(points) <- all_teams

  goals_scored <- rep(0, n_teams)
  names(goals_scored) <- all_teams

  goals_against <- rep(0, n_teams)
  names(goals_against) <- all_teams

  wins <- rep(0, n_teams)
  names(wins) <- all_teams

  draws <- rep(0, n_teams)
  names(draws) <- all_teams

  loss <- rep(0, n_teams)
  names(loss) <- all_teams


  # Loop trough all games
  for (ii in 1:ngames){

    games_played[team1[ii]] <- games_played[team1[ii]] + 1
    games_played[team2[ii]] <- games_played[team2[ii]] + 1

    goals_scored[team1[ii]] <- goals_scored[team1[ii]] + goals1[ii]
    goals_scored[team2[ii]] <- goals_scored[team2[ii]] + goals2[ii]

    goals_against[team1[ii]] <- goals_against[team1[ii]] + goals2[ii]
    goals_against[team2[ii]] <- goals_against[team2[ii]] + goals1[ii]

    if (goaldiff[ii] > 0){
      wins[team1[ii]] <- wins[team1[ii]] + 1
      loss[team2[ii]] <- loss[team2[ii]] + 1

    } else if (goaldiff[ii] == 0){
      draws[team1[ii]] <- draws[team1[ii]] + 1
      draws[team2[ii]] <- draws[team2[ii]] + 1
    } else if (goaldiff[ii] < 0) {
      wins[team2[ii]] <- wins[team2[ii]] + 1
      loss[team1[ii]] <- loss[team1[ii]] + 1

    } else {
      stop('sdfd')
    }

  }

  # TODO: Implement head-2-head tiebreaker rule.

  points = wins*3 + draws*1
  gd <- goals_scored - goals_against

  res <- data.frame(Team = all_teams,
                    Played = games_played,
                    Won = wins,
                    Drawn = draws,
                    Lost = loss,
                    Goals_scored = goals_scored,
                    Goals_conceded = goals_against,
                    Goal_difference = gd,
                    Pts = points)


  res <- res[order(res$Pts, res$Goal_difference, res$Goals_scored, decreasing = TRUE),]

  row.names(res) <- NULL

  return(res)

}



#' Scoring rules to evaluate prediction accuracy
#'
#' This function provides scoring rules for evaluating prediction accuracy.
#'
#' Currently three scoring rules are available: The log score, Brier score, and
#' the Ranked Probability Score (RPS).
#'
#' The log score is just the negative logarithm of the probability of the observed outcome.
#'
#' For all three scoring functions, a lower score means a better predictions. They will attain a
#' lowest possible score of 0 if the observed outcome were predicted with a 100% probability.'
#' The RPS has an upper limit of 1, which indicates the worst possible
#' prediction. The Brier score has an upper limit of 2. The log score has no upper
#' limit, and will be infinite if the observed outcome were predicted with probability 0.
#'
#'
#' @param predictions A matrix or data frame with probabilities, with one column for each outcome, and one row for each prediction.
#' @param observed Numeric or character vector of the same length as the number of predictions. It must contain an indicator of the observed outcome,
#' either a column number or a column name.
#' @param score Character vector of the scoring functions to use. Currently 'log', 'brier', and 'rps' are available.
#'
#' @references
#' Wheatcroft, E. (2021) Evaluating probabilistic forecasts of football matches: the case against the ranked probability score.
#' https://doi.org/10.1515/jqas-2019-0089
#'
#' Constantinou, A. C., and N. E. Fenton (2012) Solving the Problem of Inadequate Scoring Rules for Assessing Probabilistic Football Forecast Models. https://doi.org/10.1515/1559-0410.1418
#'
#'
#' @export
score_predictions <- function(predictions, observed, score){


  scorefunctions <- c('log', 'brier', 'rps')
  stopifnot(is.character(score),
            length(score) >= 1)

  score <- tolower(score)

  if (!all(score %in% scorefunctions)){
    stop('score must be one of brier, log or rps.')
  }

  # Some useful quantities.
  ncat <- ncol(predictions)
  npred <- nrow(predictions)

  stopifnot(any(observed >= 1),
            any(observed <= ncat),
            length(observed) == npred)

  # Output list.
  res <- list()

  # Expand observed vector to indicator matrix of the same
  # dimensions as the predictions matrix.
  obsmat <- matrix(0, ncol=ncat, nrow=npred)
  for (rr in 1:npred){
    obsmat[rr, observed[rr]] <- 1
  }


  # Logarithmic scoring rule
  if ('log' %in% score){
    log_scores <- numeric(npred)
    for (rr in 1:npred){
      log_scores[rr] <- -log(predictions[rr,observed[rr]])
    }

    res$log <- log_scores

  }


  if ('brier' %in% score){
    res$brier <- rowSums((predictions - obsmat)^2)
  }


  # Ranked Probability Score. (RPS).
  if ('rps' %in% score){
    rankprobscore <- numeric(npred)

    for (rr in 1:npred){
      cumulative <- 0
      for (i in 1:ncat){
        cumulative <- cumulative + (sum(predictions[rr,1:i]) - sum(obsmat[rr, 1:i]))^2
      }
      rankprobscore[rr] <- (1/(ncat-1))*cumulative
    }

    res$rps <- rankprobscore

  }

  return(res)

}





